docs/pages/impl/config.md
# Configuration

## Basic Information

Configuration is an important part of every application.

The framework provides a single point for reading the configuration through a special method that you can add to an [application](./application.md#applicationconfiginit) struct:
    ```go hl_lines="12 22"
    package application

    import (
        "github.com/componego/componego"
    )

    type Application struct {
    }

    // ...

    func (a *Application) ApplicationConfigInit(appMode componego.ApplicationMode, options any) (map[string]any, error) {
        return map[string]any{
            "config.key1": "config.value1",
            "config.key2": "config.value2",
            // ...
        }, nil
    }

    var (
        _ componego.Application           = (*Application)(nil)
        _ componego.ApplicationConfigInit = (*Application)(nil)
    )
    ```
This method returns a map containing the configuration keys and values.
You can also return an error if there was an issue reading the configuration.

This method is called only once and should return the configuration for the [application](./application.md) and [all components](./component.md) within that application.

## Configuration Reader

You can read the configuration in various ways.

For example, you can use third-party libraries to obtain the configuration.
However, your function or library must return a variable of the type ^^map[string]any^^:
    ```go hl_lines="7 18 25"
    package application

    import (
        "fmt"

        "github.com/componego/componego"
        "github.com/spf13/viper"
    )

    type Application struct {
    }

    // ...

    func (a *Application) ApplicationConfigInit(appMode componego.ApplicationMode, options any) (map[string]any, error) {
        switch appMode {
        case componego.ProductionMode:
            return ConfigReader("./config/production.config.json")
        // ...
        default:
            return nil, fmt.Errorf("not supported application mode: %d", appMode)
        }
    }

    func ConfigReader(filename string) (map[string]any, error) {
        // This function should be placed in a separate package.
        v := viper.New()
        v.SetConfigFile(filename)
        if err := v.ReadInConfig(); err != nil {
            return nil, err
        }
        return v.AllSettings(), nil
    }

    var (
        _ componego.Application           = (*Application)(nil)
        _ componego.ApplicationConfigInit = (*Application)(nil)
    )
    ```

You can also perform post-processing of values after reading the configuration:
    ```go hl_lines="5 13"
    package config

    import (
        "github.com/componego/componego/impl/environment/managers/config"
        "github.com/spf13/viper"
    )

    func ConfigReader(filename string) (map[string]any, error) {
        // This function should be placed in a separate package.
        v := viper.NewWithOptions()
        // ...
        settings := v.AllSettings()
        return settings, config.ProcessVariables(settings)
    }

    ```

This function converts the following values:
    ```json
    {
        "server": {
            "addr": ":${ENV:APP_PORT}"
        }
    }
    ```
into:
    ```json
    {
        "server": {
            "addr": ":1234"
        }
    }
    ```
You can also use the default value after a pipe:
    ```json
    {
        "server": {
            "addr": ":${ENV:APP_PORT|1234}"
        }
    }
    ```

We have described the map that ^^ApplicationConfigInit^^ returns.
The next section describes how to get this value in your [application](./application.md) or [component](./component.md).

## Configuration Getter

Any configuration values can be accessed using the [environment](./environment.md) in any part of the application:
    ```go
    value, err := env.ConfigProvider().ConfigValue("server.addr", nil)
    ```

Golang is a strongly typed language and it is impossible to use generics ^^in this case^^ in the current version of the language (go1.22).

There is an additional function within the framework that helps solve the typing problem:
    ```go hl_lines="5 9"
    package config

    import (
        "github.com/componego/componego"
        "github.com/componego/componego/impl/environment/managers/config"
    )

    func GetServerAddr(env componego.Environment) (string, error) {
        return config.Get[string]("server.addr", nil, env)
    }
    ```
The shorter code appears as follows:
    ```go hl_lines="5 9"
    package config

    import (
        "github.com/componego/componego"
        "github.com/componego/componego/impl/environment/managers/config"
    )

    func GetServerAddr(env componego.Environment) string {
        return config.GetOrPanic[string]("server.addr", nil, env)
    }
    ```

!!! note
    We use a dot as a separator between configuration keys to indicate different levels of nesting.

## Configuration Processor

Validation and transformation of configuration values can be performed using [processors](./processor.md):
    ```go hl_lines="6"
    package config

    import (
        "github.com/componego/componego"
        "github.com/componego/componego/impl/environment/managers/config"
        "github.com/componego/componego/impl/processors"
    )

    func GetServerAddr(env componego.Environment) string {
        return config.GetOrPanic[string]("server.addr", processors.Multi(
            processors.DefaultValue(":3030"),
            processors.ToString(),
        ), env)
    }
    ```

!!! note
    Typing is a crucial aspect of processors. The generic type must match the processor. Otherwise, there may be an error:
    ```go hl_lines="3"
    config.GetOrPanic[int64]("server.port", processors.Multi(
        processors.IsRequired(),
        processors.ToInt64(),
    ), env)
    ```
    You should use a processor to change the type, as there is no guarantee that ^^ApplicationConfigInit^^ will return a value of the desired type.

## Configuration Struct

Application configurations can indeed become quite large, and managing each configuration key with individual [processors](./processor.md) can be inefficient.
Instead, it's recommended to define a struct that contains fields corresponding to different configuration values.
You can validate this struct in a single step during the initialization of the dependency injection [DI container](./dependency.md).

Each component can have its own separate struct to describe its configuration.
This allows for better modularity and separation of concerns.
You can find an example of such a struct-based configuration approach [here](https://medium.com/@konstanchuk/25bfd16a97a9#413f){:target="_blank"}.

## Configuration Examples

It’s recommended to create an example configuration file when you create an [application](./application.md) or [component](./component.md).

For instance, if you create a component, you should also include a file with an example of its configuration.
This allows developers who use the component to easily copy and merge the example configuration file into their main application configuration file.
After that, it can be read in the ^^ApplicationConfigInit^^ function.


docs/pages/impl/environment.md
# Application Environment

## Basic information

The environment package serves as a comprehensive solution for managing an application runtime.
This package not only provides access to [dependency injection (DI)](./dependency.md) management but also includes features for
handling application [active components](./component.md), [configuration](./config.md), [application mode](./runner.md#application-mode),
[IO](./environment.md#application-io) and [global context](./environment.md#application-context).

## How to get environment

The environment can be accessed in several ways.
The first and easiest way is to get this object in the [application action](./application.md#applicationaction):
    ```go
    func (a *Application) ApplicationAction(env componego.Environment, _ []string) (int, error) {
        env.GetContext()
        env.Application()
        env.ApplicationIO()
        env.ApplicationMode()
        env.Components()
        env.DependencyInvoker()
        env.ConfigProvider()

        // ...
        return componego.SuccessExitCode, nil
    }

    var _ componego.Application = (*Application)(nil)
    ```

You can also get this object via DI:
    ```go
    type MyType struct {
        env componego.Environment `componego:"inject"`
    }

    func (m *MyType) Method() {
        if m.env.ApplicationMode() == componego.DeveloperMode {
            // ...
        }
    }
    ```
or
    ```go
    err := dependencyInvoker.Invoke(func(env componego.Environment) {
        // ...
    })
    ```

or another ways described in the [documentation about DI](./dependency.md).

!!! note
    The environment object cannot be rewritten by [rewritten DI objects](./dependency.md#rewriting-dependencies).
    This object is present in any application.


## How to use environment

| Method                  | Description                                                | Documentation                                  |
|-------------------------|------------------------------------------------------------|------------------------------------------------|
| env.GetContext()        | returns a current application context                      | [open](./environment.md#application-context)   |
| env.SetContext(newCtx)  | sets a new application context                             | [open](./environment.md#application-context)   |
| env.Application()       | returns a current application object                       | [open](./application.md)                       |
| env.ApplicationIO()     | returns an object for getting application input and output | [open](./environment.md#application-io)        |
| env.ApplicationMode()   | returns the mode in which the application is started       | [open](./runner.md#application-mode)           |
| env.Components()        | returns a sorted list of active application components     | [open](./component.md)                         |
| env.DependencyInvoker() | returns an object to invoke dependencies                   | [open](./dependency.md#access-to-dependencies) |
| env.ConfigProvider()    | returns an object for getting config                       | [open](./config.md#configuration-getter)       |


This serves as a universal key for accessing any part of the application.

## Application Context

It is recommended to use the application context to run various functions.
You can also replace the current context with another one, but the new context must inherit from the previous main context:
    ```go hl_lines="14 16"
    package component

    import (
        "context"
        "time"

        "github.com/componego/componego"
        "github.com/componego/componego/impl/environment/managers/component"
    )

    func NewComponent() componego.Component {
        factory := component.NewFactory("example", "0.0.1")
        factory.SetComponentInit(func(env componego.Environment) error {
            ctx, cancelCtx := context.WithTimeout(env.GetContext(), time.Second*100)
            // ...
            return env.SetContext(ctx)
        })
        return factory.Build()
    }
    ```

## Application IO

If you want to output (or receive) some text to (from) the console, you must use special methods:
    ```go hl_lines="2"
    func (a *Application) ApplicationAction(env componego.Environment, _ []string) (int, error) {
        appIO := env.ApplicationIO()
        _, _ = fmt.Fprintln(appIO.OutputWriter(), "your text")
        _, _ = fmt.Fprintln(appIO.ErrorOutputWriter(), "your error text")
        reader := bufio.NewReader(appIO.InputReader())
        text, _ := reader.ReadString('\n')
    }
    ```


CHANGELOG.md
## 0.2.0 (September 26, 2024)

* Improved documentation and testing.
* Fixed some bugs.

## 0.1.0 (July 27, 2024)

* All interfaces are stable now and will no longer be changed.
* Improved the graceful shutdown.
* Fixed dependency closing order.
* Increased GitHub Action speed.

## 0.0.3 (July 8, 2024)

* The graceful shutdown component was replaced with various application launch functions.
* Added the ability to conveniently close resources in dependencies.
* Improved the xerrors package and added error codes.
* Added strict code formatting.

## 0.0.2 (June 30, 2024)

* Increased test coverage.
* Improved dependency injection stability.
* Improved several interfaces for flexibility.
* Added linking of application skeleton creation to the last git tag.

## 0.0.1 (June 1, 2024)

* Initial release.


examples/url-shortener-app/pkg/components/server/internal/config.go
package internal

import (
	"time"

	"github.com/componego/componego"
	"github.com/componego/componego/impl/environment/managers/config"
	"github.com/componego/componego/impl/processors"
)

type Config struct {
	Addr              string
	ReadTimeout       time.Duration
	ReadHeaderTimeout time.Duration
	WriteTimeout      time.Duration
	IdleTimeout       time.Duration
	StopTimeout       time.Duration
}

func NewConfig(env componego.Environment) *Config {
	return &Config{
		Addr: config.GetOrPanic[string]("server.addr", processors.Multi(
			processors.DefaultValue(":3030"),
			processors.ToString(),
		), env),
		ReadTimeout:       1 * time.Second,
		ReadHeaderTimeout: 1 * time.Second,
		WriteTimeout:      1 * time.Second,
		IdleTimeout:       30 * time.Second,
		StopTimeout:       10 * time.Second,
	}
}
